preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp

```diff
@@ -0,0 +1,254 @@
+    BOOL hasCurrent = FALSE;
+    RETURN_IF_FAILED(firewallRuleEnum->GetHasCurrent(&hasCurrent));
+    while (hasCurrent)
+    {
+        FirewallRule firewallRule;
+        ComPtr<IMsixElement> ruleElement;
+        RETURN_IF_FAILED(firewallRuleEnum->GetCurrent(&ruleElement));
+
+        Text<wchar_t> direction;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(firewallRuleDirectionAttribute.c_str(), &direction));
+        firewallRule.direction = direction.Get();
+
+        Text<wchar_t> protocol;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(protocolAttribute.c_str(), &protocol));
+        firewallRule.protocol = protocol.Get();
+
+        Text<wchar_t> profile;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(profileAttribute.c_str(), &profile));
+        firewallRule.profile = profile.Get();
+
+        Text<wchar_t> localPortMin, localPortMax, remotePortMin, remotePortMax;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(localPortMinAttribute.c_str(), &localPortMin));
+        if (localPortMin.Get() != nullptr)
+        {
+            firewallRule.localPortMin = localPortMin.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(localPortMaxAttribute.c_str(), &localPortMax));
+        if (localPortMax.Get() != nullptr)
+        {
+            firewallRule.localPortMax = localPortMax.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(remotePortMinAttribute.c_str(), &remotePortMin));
+        if (remotePortMin.Get() != nullptr)
+        {
+            firewallRule.remotePortMin = remotePortMin.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(remotePortMaxAttribute.c_str(), &remotePortMax));
+        if (remotePortMax.Get() != nullptr)
+        {
+            firewallRule.remotePortMax = remotePortMax.Get();
+        }
+
+        m_firewallRules.push_back(firewallRule);
+
+        RETURN_IF_FAILED(firewallRuleEnum->MoveNext(&hasCurrent));
+    }
+
+    return S_OK;
+}
+
+HRESULT FirewallRules::AddFirewallRules(FirewallRule& firewallRule)
+{
+    INetFwPolicy2 *pNetFwPolicy2 = NULL;
+    INetFwRules *pFwRules = NULL;
+    INetFwRule *pFwRule = NULL;
+
+    // Retrieve INetFwPolicy2
+    RETURN_IF_FAILED(WFCOMInitialize(&pNetFwPolicy2));
+
+    // Retrieve INetFwRules
+    RETURN_IF_FAILED(pNetFwPolicy2->get_Rules(&pFwRules));
+
+    // Create a new Firewall Rule object.
+    RETURN_IF_FAILED(CoCreateInstance(__uuidof(NetFwRule), NULL, CLSCTX_INPROC_SERVER, __uuidof(INetFwRule), (void**)&pFwRule));
+
+    std::wstring packageDisplayName = m_msixRequest->GetPackageInfo()->GetDisplayName();
+    BSTR ruleName = SysAllocString(packageDisplayName.data());
+    BSTR ruleDescription = SysAllocString(packageDisplayName.data());
+    pFwRule->put_Name(ruleName);
+    pFwRule->put_Description(ruleDescription);
+    pFwRule->put_Enabled(VARIANT_TRUE);
+

```

Comment:
these return an HRESULT, wrap them in RETURN_IF_FAILED as well.

Reply:
Used RETURN_IF_FAILED.